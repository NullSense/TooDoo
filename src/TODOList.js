import React from 'react';
import TODOItem from './TODOItem.js';

/**
 * TODOList has top level state in which all TODOItems are stored
 * @version 0.1
 * @author salsa20
 * @extends React.Component
 */
class TODOList extends React.Component {
  constructor() {
    super();

    // 'global' state
    this.state = {
      items: [{ id: Date.now(), entry: '' }]
    };
  }

  /**
   * add an item to global state
   */
  addItem() {
    this.setState(prev => {
      return { items: [...prev.items, { id: Date.now() }] };
    });
  }

  /**
   * delete an item from global state
   */
  delItem(id) {
    this.setState(prev => {
      return { items: prev.items.filter(entry => entry.id !== id) };
    });
  }

  /**
   * load Item through API call
   */
  loadItems() {
    fetch('https://jsonplaceholder.typicode.com/todos/1') // use mock api for now
      .then(response => response.json())
      .then(json => {
        // check if item with this id is already contained
        if (this.state.items.findIndex(item => item.id === json.id) === -1) {
          this.setState({ items: [...this.state.items, { id: json.id, entry: json.title }] });
        }
      });
  }

  render() {
    return (
      <div className="TODOList">
        {this.state.items.map(item => (
          <TODOItem key={item.id} id={item.id} entry={item.entry} del={this.delItem.bind(this)} />
        ))}
        <div className="container">
          <button className="Button" type="button" onClick={this.addItem.bind(this)}>
            +
          </button>
          <button style={{ width: '200px' }} className="Button" type="button" onClick={this.loadItems.bind(this)}>
            Load
          </button>
        </div>
      </div>
    );
  }
}

export default TODOList;
