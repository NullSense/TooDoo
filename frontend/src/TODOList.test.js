import { configure, shallow } from 'enzyme';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import TODOList from './TODOList.js';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });

describe('TODOList', () => {
  // always setup a component with empty props
  let component;
  beforeEach(() => {
    component = shallow(<TODOList />);
  });

  // unmount after each test
  // afterEach(() => {
  //   component.unmount();
  // });

  /*
   * snapshot testing
   */
  it('Should render correctly', () => {
    const snapshot = renderer.create(<TODOList />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  // there should be only one TODOList
  it('should have exactly one TODOList', () => {
    const wrapper = component.find('.TODOList');
    expect(wrapper.length).toEqual(1);
  });

  // there should be only one TODOList
  it('should initially hold one item', () => {
    expect(component.state().items.length).toEqual(1);
  });

  // delItem should work as expected on call
  it('should delete item from global state', () => {
    component.setState({
      items: [{ id: 0 }]
    });
    expect(component.state().items.length).toEqual(1);
    component.instance().delItem(0);
    expect(component.state().items.length).toEqual(0);
  });

  // addItem should work as expected on call
  it('should add item to global state', () => {
    component.setState({
      items: []
    });
    expect(component.state().items.length).toEqual(0);
    component.instance().addItem({ preventDefault() {}, value: '' });
    expect(component.state().items.length).toEqual(1);
    expect(component.state().items[0].id).toBeLessThanOrEqual(Date.now());
  });

  // addItem should work as expected on submit
  it('should add item on submit', () => {
    const prevLength = component.state().items.length;
    component
      .find('button')
      .at(0)
      .simulate('click', { preventDefault() {} });
    expect(component.state().items.length).toEqual(prevLength + 1);
  });

  // loadItem should correctly make an api call
  // TODO: mock api call
  it('should make a proper api call', async () => {
    component.setState({
      items: []
    });
    expect(component.state().items.length).toEqual(0);
    await component.instance().loadItems();
    expect(component.state().items.length).toEqual(1);
    expect(component.state().items[0]).toEqual({ id: 1, entry: 'delectus aut autem' });
  });
  // it('should make an api call on button click', async () => {
  //   component.setState({
  //     items: []
  //   });
  //   expect(component.state().items.length).toEqual(0);
  //   component
  //     .find('button')
  //     .at(1)
  //     .simulate('click');
  //   expect(component.state().items.length).toEqual(1); // this is called before the async returns....
  //   expect(component.state().items[0]).toEqual({ id: 1, entry: 'delectus aut autem' });
  // });
});
