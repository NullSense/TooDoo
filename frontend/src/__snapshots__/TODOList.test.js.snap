// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TODOList Should render correctly 1`] = `
<div
  className="TODOList"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="EntryWrapper"
    >
      <li
        className="TODOItem"
        onClick={[Function]}
      >
        
      </li>
      <button
        className="DelEntry"
        type="submit"
      >
        <img
          alt="-"
          src="https://img.icons8.com/metro/26/000000/trash.png"
          style={
            Object {
              "filter": "opacity(50%)",
              "height": "auto",
              "maxWidth": "50%",
            }
          }
        />
      </button>
    </div>
  </form>
  <div
    className="container"
  >
    <button
      className="Button"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      className="Button"
      onClick={[Function]}
      style={
        Object {
          "width": "200px",
        }
      }
      type="button"
    >
      Load
    </button>
  </div>
</div>
`;
